### PR TITLE
Add error info to quarantine object (closes #6722)

### LINKED ISSUE #6722: Provide more information about errors for each test run in quarantine mode
### What is your Scenario?

In general, we need information about test runs for each quarantine mode attempt in each browser:

* the status of a quarantine test run in a specific browser (successful/failed)
* information about a JS client-side error that occurs on each test run

 I need to process this information with a reporter and display it in the Dashboard

### What are you suggesting?

- Support an additional reporter method `reportTestRunDone`
- Give more verbose data in the `reportTestDone`

### What alternatives have you considered?

_No response_

### Additional context

https://teams.microsoft.com/l/message/19:7e845d3e5e324502abae3417a2ac8c7c@thread.skype/1637681369524?tenantId=e4d60396-9352-4ae8-b84c-e69244584fa4&groupId=c3abb0f3-231c-4d87-b5a3-44925dc1d075&parentMessageId=1637681369524&teamName=TestCafe&channelName=Framework%20(Open%20source)&createdTime=1637681369524

### PR DESCRIPTION
**Pull request by Artem-Babich**

### CODE REVIEW

File: test/functional/fixtures/regression/gh-6722/test.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            };
+        },
+        output: {
+            write: () => {
+            },
+            end: () => {
+            },
+        },
+    };
+};
```
Comment (AlexKamaev) [inline test/functional/fixtures/regression/gh-6722/test.js:31]: Check this util
...\testcafe\test\functional\utils\reporter.js

File: test/functional/fixtures/regression/gh-6722/test.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        return runTests('testcafe-fixtures/index.js', 'Paste text on button click', {
+            reporter,
+            skipJsErrors:   false,
+            quarantineMode: true,
+        }).then(() => {
+            expect(result.quarantine[1].passed).to.equal(true) &&
+            expect(result.quarantine[2]).to.be.undefined;
+        });
+    });
+
```
Comment (AlexKamaev) [inline test/functional/fixtures/regression/gh-6722/test.js:64]: delete this test

File: test/functional/fixtures/regression/gh-6722/testcafe-fixtures/index.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            await t
+                .typeText('#testInput', `Custom client exception on test #${ testNumber }`)
+                .click('#failButton')
+                .expect(Selector('#testInput').value).eql('success');
+            break;
+        case ERRORS.None:
+            await t
+                .click('#successButton')
+                .expect(Selector('#testInput').value).eql('success');
+            break;
```
Comment (AlexKamaev) [inline test/functional/fixtures/regression/gh-6722/testcafe-fixtures/index.js:22]: Maybe we do not need the Errors.None case?

File: test/functional/fixtures/regression/gh-6722/testcafe-fixtures/index.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+fixture`GH-1994 - The element that matches the specified selector is not visible`
+    .page`http://localhost:3000/fixtures/regression/gh-6722/pages/index.html`;
+
+test(`Paste text on button click`, async t => {
+    await t
+        .click('#successButton')
+        .expect(Selector('#testInput').value).eql('success');
+});
```
Comment (AlexKamaev) [inline test/functional/fixtures/regression/gh-6722/testcafe-fixtures/index.js:33]: delete

File: src/reporter/index.ts [hunk 5]
```diff
@@ -362,14 +367,18 @@ export default class Reporter {
...
+            testItem.quarantine = testRun.quarantine.attempts.reduce((result: Record<string, object>, { errors }, index: number) => {
                 const passed            = !errors.length;
                 const quarantineAttempt = index + 1;
 
                 result[quarantineAttempt] = { passed };
 
                 return result;
-            }, {});
+
+            }, testItem.quarantine);
```
Comment (AlexKamaev) [inline src/reporter/index.ts:381]: Let use Object.assign instead on initializing reduce this testItem.quarantine

**Review by AlexKamaev — CHANGES_REQUESTED** [review]

File: src/reporter/index.ts [hunk 6]
```diff
@@ -362,14 +367,17 @@ export default class Reporter {
...
-            testItem.quarantine = testRun.quarantine.attempts.reduce((result: Record<string, object>, errors: TestRunErrorFormattableAdapter[], index: number) => {
+            const testItemQuarantine = testRun.quarantine.attempts.reduce((result: Record<string, object>, { errors }, index: number) => {
                 const passed            = !errors.length;
                 const quarantineAttempt = index + 1;
 
                 result[quarantineAttempt] = { passed };
 
                 return result;
-            }, {});
+
```
Comment (AlexKamaev) [inline src/reporter/index.ts:377]: excessive empty line

File: src/reporter/index.ts [hunk 7]
```diff
@@ -501,13 +509,32 @@ export default class Reporter {
...
 
-        reportItem.browsers.push(Object.assign({ testRunId: testRun.id }, testRun.browser));
+        if (testRun.quarantine) {
+            reportItem.quarantine = reportItem.quarantine || {};
+
+            const reportItemQuarantine = testRun.quarantine.attempts.reduce((result: Record<string, object>, { errors, testRunId }) => {
+                const passed = !errors.length;
+
+                if (!browser.quarantineAttemptsTestRunIds)
+                    browser.quarantineAttemptsTestRunIds = [];
```
Comment (AlexKamaev) [inline src/reporter/index.ts:528]: browser.quarantineAttemptsTestRunIds = browser.quarantineAttemptsTestRunIds || []

File: src/reporter/index.ts [hunk 8]
```diff
@@ -501,13 +509,32 @@ export default class Reporter {
...
+
+                browser.quarantineAttemptsTestRunIds.push(testRunId);
+                result[testRunId] = { passed, errors };
```
Comment (AlexKamaev) [inline src/reporter/index.ts:531]: add empty line before assignment

File: test/functional/fixtures/regression/gh-6722/constants.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    ERRORS.Client,
+    ERRORS.None,
+    ERRORS.None,
+    ERRORS.Server,
+];
+
+const Counter = function () {
+    this.counters = {};
+
+    this.add = (alias) => {
```
Comment (AlexKamaev) [inline test/functional/fixtures/regression/gh-6722/constants.js:26]: no need in `(` and `)` around alias

File: test/functional/fixtures/regression/gh-6722/test.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    const currentSuccessAttemptsCount = Object.values(quarantine).filter(attempt => attempt.passed).length;
+    const currentFailedAttemptsCount  = Object.values(quarantine).filter(attempt => !attempt.passed).length;
+    const currentTestRunIdsCount      = browsers.reduce((counter, browser) => counter + browser.quarantineAttemptsTestRunIds.length, 0);
+
+    // We have to add the number of attempts per browser as we also kept the old behavior. {1:{}, 2:{},  ... testRunId1:{}, testRunId2:{}, ...}
+    const expectedAttemptsCount        = attemptsCount * (browsersCount + 1);
+    const expectedSuccessAttemptsCount = successAttemptsCount * (browsersCount + 1);
+    const expectedFailedAttemptsCount  = failedAttemptsCount * (browsersCount + 1);
+
+    return expect(currentAttemptsCount).to.equal(expectedAttemptsCount) &&
```
Comment (AlexKamaev) [inline test/functional/fixtures/regression/gh-6722/test.js:33]: No need to return anything from this function.
No need to add && between conditions.
It's better to write assertions line by line:
```
        expect(currentSuccessAttemptsCount).to.equal(expectedSuccessAttemptsCount);
        expect(currentFailedAttemptsCount).to.equal(expectedFailedAttemptsCount);
        expect(currentTestRunIdsCount).to.equal(attemptsCount * browsersCount);
```

**Review by AlexKamaev — CHANGES_REQUESTED** [review]

**Review by AlexKamaev — APPROVED** [review]

**Review by Dmitry-Ostashev — APPROVED** [review]

**Pull request merged**